aten/src/ATen/cuda/CUDAStream.cpp

```diff
@@ -60,6 +62,96 @@ static std::deque<std::atomic<uint32_t>> high_priority_counters;
 static std::vector<std::array<CUDAStreamInternals, kStreamsPerPool>> low_priority_streams;
 static std::vector<std::array<CUDAStreamInternals, kStreamsPerPool>> high_priority_streams;
 
+// Note [StreamId assignment]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~
+// How do we assign stream IDs?
+//
+// -- 2 bits --  -- 5 bits -----
+// StreamIdType  stream id index
+//  00 = default stream
+//  01 = low priority stream
+//  10 = high priority stream
+//
+// This is not really for efficiency; it's just easier to write the code
+// to extract the index if we do this with bitmasks :)
+//
+// This is entirely an internal implementation detail, we reserve the right to
+// renumber streams however we like.
+
+enum class StreamIdType : uint8_t {
+  DEFAULT = 0x0,
+  LOW     = 0x1,
+  HIGH    = 0x2,
+};
+
+std::ostream& operator<<(std::ostream& stream, StreamIdType s) {
+  switch (s) {
+    case StreamIdType::DEFAULT:
+      stream << "DEFAULT";
+      break;
+    case StreamIdType::LOW:
+      stream << "LOW";
+      break;
+    case StreamIdType::HIGH:
+      stream << "HIGH";
+      break;
+    default:
+      stream << static_cast<uint8_t>(s);
+      break;
+  }
+  return stream;
+}
+
+// Promotion makes me nervous...
+
+static inline StreamIdType streamIdType(StreamId s) {
+  return static_cast<StreamIdType>(s >> kStreamsPerPoolBits);
+}
+
+static inline size_t streamIdIndex(StreamId s) {
+  return static_cast<size_t>(s & ((1 << kStreamsPerPoolBits) - 1));
+}
+
+StreamId makeStreamId(StreamIdType st, size_t si) {
+  return (static_cast<StreamId>(st) << kStreamsPerPoolBits) | static_cast<StreamId>(si);
+}
+
+template <typename T>
+static bool pointer_within(const T* ptr, ArrayRef<T> arr) {
+  return std::greater_equal<const T*>()(ptr, arr.data()) && std::less<const T*>()(ptr, arr.data() + arr.size());
+}
+
+static StreamId CUDAStream_getStreamId(const CUDAStreamInternals* ptr) {
```

Comment:
Yes, the idea is we only ever pass around Stream. PyTorch is pretty close to this, since we're passing around a stream object of some sort, and Caffe2 this would just be a codemod change, since the representation is isomorphic to the current Caffe2 representation (which is also an int.)

Storing a StreamId in CUDAStreamInternals is definitely less magical, but not by much, because you still have to write the code that maps from a StreamId to an appropriate stream, and the most convenient way to do that is some sort of bitmasky thing. But we could certainly store StreamId in the stream, on top of doing this.